tds.go

```diff
@@ -465,6 +465,134 @@ func sendLogin(w *TdsBuffer, login login) error {
 	return w.FinishPacket()
 }
 
+// Login embeds login. This is a hack to expose login to the outside world while minimising
+// changes to code.
+type Login struct {
+	login
+}
+
+// offsetAfterHeader calculates an offset adjusted for reading from a buffer that starts
+// after the packet header.
+func offsetAfterHeader(offset uint16) int {
+	return headerSize + int(offset)
+}
+
+// ReadLogin parses a TDS7 login packet.
+func ReadLogin(r *TdsBuffer) (*Login, error) {
+	var err error
+
+	packet_type, err := r.BeginRead()
+	if err != nil {
+		return nil, err
+	}
+
+	if packet_type != PackLogin7 {
+		return nil, errors.New(fmt.Sprintf("Invalid response, expected packet type %d", PackLogin7))
+	}
+
+	hdr := loginHeader{}
+
+	err = binary.Read(r, binary.LittleEndian, &hdr)
+	if err != nil {
+		return nil, err
+	}
+
+	r.rpos = offsetAfterHeader(hdr.HostNameOffset)
```

Comment:
*Probably* not worth it, but you could possibly eliminate most of the `r.rpos = offsetAfterHeader(***)` lines by creating a helper function:
```
func readUcs2WithOffset(r io.Reader, numchars int, offset uint16) (res string, err error) {
    r.pos = offsetAfterHeader(offset)
    return readUcs2(r, numchars)
}
```
where the calls would be e.g.:
```
    hostname, err := readUcs2WithOffset(r, int(hdr.HostnameLength), hdr.HostNameOffset)
```
But it's 6 o' one, half-dozen of another.

Reply:
@diverdane Correction: _Definitely_ worth it :)